server/src/main/java/org/elasticsearch/script/field/NestedDocument.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.script.field;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+public class NestedDocument {
+    private final WriteField parent;
+    private final Map<String, Object> doc;
+
+    public NestedDocument(WriteField parent, Map<String, Object> doc) {
+        this.parent = parent;
+        this.doc = Objects.requireNonNull(doc);
+    }
+
+    /**
+     * Get a {@link WriteField} inside this NestedDocument, all operation on the {@link WriteField} are relative to this NestedDocument.
+     */
+    public WriteField field(String path) {
+        return new WriteField(path, this::getDoc);
+    }
```

Comment:
Instead of forcing WriteField to take in a supplier, would it make more sense to just have another constructor take in a reference?